[RISCV][TTI] Implement instruction cost for vp.reduce.*

ElvisWang123 opened the pull request:
This patch based on #114180.

ElvisWang123 committed:
Precommit testcases.

ElvisWang123 committed:
[RISCV][TTI] Implement instruction costs for vp.reduce.*.

ElvisWang123 committed:
Fix missing tests

preames commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1179,6 +1179,79 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
     return getCmpSelInstrCost(Instruction::Select, ICA.getReturnType(),
                               ICA.getArgTypes()[0], CmpInst::BAD_ICMP_PREDICATE,
                               CostKind);
+  case Intrinsic::vp_reduce_add:
```
Please see CachingVPExpander::expandPredicationInReduction as an example of how to perform these mappings more cleanly.

ElvisWang123 replied:
Updated, thanks!

preames replied:
Please use:

```
Intrinsic::ID RedID = *VPI.getFunctionalIntrinsicID();
unsigned Opc = getArithmeticReductionInstruction(RedID);
OR
getMinMaxReductionIntrinsicOp(RedID);
```

ElvisWang123 replied:
Fixed, thanks!!

ElvisWang123 committed:
Fixup! early return

ElvisWang123 committed:
Fixup! using helper function to get the reduction opcode.

preames commented on llvm/test/Analysis/CostModel/RISCV/reduce-add.ll:
```diff
@@ -1,8 +1,10 @@
-; RUN: opt < %s -mtriple=riscv64 -mattr=+v -passes="print<cost-model>" -cost-kind=throughput 2>&1 -disable-output | FileCheck %s
-; RUN: opt < %s -mtriple=riscv32 -mattr=+v -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output | FileCheck %s --check-prefix=SIZE
-; RUN: opt < %s -mtriple=riscv64 -mattr=+v -passes="print<cost-model>" -cost-kind=code-size 2>&1 -disable-output | FileCheck %s --check-prefix=SIZE
+; RUN: opt < %s -mtriple=riscv32 -mattr=+v -passes="print<cost-model>" -cost-kind=throughput 2>&1 -disable-output --type-based-intrinsic-cost=true | FileCheck %s
```
Please revert the change (in this and other files) to add --type-based-intrinsic-cost=true.

ElvisWang123 replied:
You mean remove `--type-based-intrinsic-cost=true` from the original `RUN` and add another version of the `RUN` line with `--type-based-intrinsic-cost=true`?  Or you thought we have better approach to test the instruction cost of vp-intrinsics here?
I am not  sure what you want?

preames replied:
You modified the run lines, please remove the modification in this change.  If you want to modify them, do that in it's own change.  (I am skeptical of your motivation for that separate change.)

ElvisWang123 replied:
Reverted. 
Add new run lines for the `type-based` queries and check the instruction costs are same as normal query.

preames reviewed (changes requested):
I am not sure the fully duplicated test coverage for vp.reduce is justified here.  The whole point of this change is that they're cost just like normal reduces?  Why do we need to complete double all test coverage?

ElvisWang123 commented:
> I am not sure the fully duplicated test coverage for vp.reduce is justified here. The whole point of this change is that they're cost just like normal reduces? Why do we need to complete double all test coverage?

The test cases here want to make sure the instruction costs for vp intrinsics are same as their non-vp counterpart.

Adding `type-based-intrinsic-cost=true` let the printer class query the cost model without values which align to how loop vectorizer query the instruction costs.

ElvisWang123 committed:
Fixup! Revert changes of original run lines and add new runs for type-based query.

The new RUNs check the instruction costs from type-based queries are same
as normal queries.

lukel97 commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -1191,6 +1191,40 @@ RISCVTTIImpl::getIntrinsicInstrCost(const IntrinsicCostAttributes &ICA,
+                                        cast<VectorType>(ICA.getArgTypes()[1]),
+                                        ICA.getFlags(), CostKind);
+    return getArithmeticReductionCost(
+        RedOp, cast<VectorType>(ICA.getArgTypes()[1]), std::nullopt, CostKind);
```
Why can't we pass ICA.getFlags() for the non FP reductions?

ElvisWang123 replied:
Fixed.
Before #114180 landed , passing FMF flags into `getArightmeicReductionCost()` with int reductions will always return the cost of `oredered fadd reduction`.

ElvisWang123 committed:
Fixup! Passing FMF flags.

ElvisWang123 committed:
Fixup! drop all RUN line changes

ElvisWang123 commented:
> I've made specific review suggestions you have not applied. As a means to make forward progress here, are you okay with me landing a modified version of this patch? The code is fine, it's only the testing bits we're stuck on. I'll simply make the testing change I've requested, and you can revisit testing improvements in a separate review if you want.

I push a new commit [2c0f6f1](https://github.com/llvm/llvm-project/pull/114184/commits/2c0f6f10a201846ceac19355ad2641269cb192d6) to revert all RUN line changes.

If the latest change still can not fit your requirement, feel free to land a modified version of this patch.

LiqinWeng commented:
why close it~

lukel97 commented:
> why close it~

It was landed separately in 3431d133ccfa75d16964be455238e2a1da0c2004
